src/EditorFeatures/Test2/Diagnostics/DiagnosticServiceTests.vb

```diff
@@ -2519,5 +2520,139 @@ class MyClass
                                                 End Sub, OperationKind.VariableDeclaration)
             End Sub
         End Class
+
+        <WpfFact, WorkItem(66968, "https://github.com/dotnet/roslyn/issues/66968")>
+        Public Async Function TestDiagnosticsForSpanDoesNotAnalyzeOutsideSpanAsync() As Task
```

Comment:
❓ Do we already have a test for a external (does not implement `IBuiltInAnalyzer`) analyzer that reports diagnostics in a different method? For example:

```csharp
class T
{
  void M1()
  {
    return; // When analyzing _this return statement_, the analyzer reports a diagnostic for _yield return_ elsewhere
  }

  IEnumerable<int> M2()
  {
    yield return 0;
  }
}
```

Reply:
> In this case, it's critical that a request for diagnostics on the yield return span execute the analyzer over the return statement, or else it will not locate the diagnostic

We explicitly do not support such incorrectly written analyzers. For example, if the method M1 was defined in a different partial type in a different file, the analyzer would not work. Such an analyzer should be implemented as a SymbolStart/End analyzer to work correctly.

Theoretically, an analyzer can report a diagnostic in any part of compilation for a callback made for any other part of compilation. If the diagnostic is reported outside the location scope associated with the callback, we classify these as non-local compilation diagnostics, for which we have never supported code fixes. These diagnostics are equivalent to compilation end diagnostics. The core analyzer executor itself would mark such diagnostics reported outside the syntax node corresponding to the original callback as non-local diagnostics: https://github.com/dotnet/roslyn/blob/aff2a21bbf6726f0ecab4660568e068aa923ef55/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs#L1339-L1386

Reply:
> For example, if the method M1 was defined in a different partial type in a different file,

I'm only talking about the case where the analyzer is reporting a diagnostic at a different location in the _same file_. My understanding is this is the only meaning behind the `SupportsSpanBasedSemanticDiagnosticAnalysis` flag, so why not just force that option for every analyzer?

Reply:
FYI, the doc comments on the field also indicates this enum was only intended for incremental method body analysis scenarios: https://github.com/dotnet/roslyn/blob/aff2a21bbf6726f0ecab4660568e068aa923ef55/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Diagnostics/DiagnosticAnalyzerCategory.cs#L9-L44